syntax = "proto3";

package policy_app;

// ==========================================
// SHARED MESSAGES
// ==========================================

// Shared context chunk structure used by both RAG and Chat services
message ContextChunk {
  string text = 1;
  string doc_id = 2;
  float score = 3;
}

// ==========================================
// 1. LLM SERVICE (The Brain)
// Used by: Chat Service
// ==========================================
service LLMService {
  // Generates text based on prompt (and context)
  rpc GenerateResponse (LLMRequest) returns (LLMResponse);
  
  // Optional: If you want streaming text generation
  rpc StreamResponse (LLMRequest) returns (stream LLMResponse);
}

message LLMRequest {
  string system_prompt = 1;
  string user_query = 2;
  string context = 3;
  float temperature = 4;
}

message LLMResponse {
  string text = 1;
}

// ==========================================
// 2. RAG SERVICE (The Knowledge Base)
// Used by: Chat Service & API Gateway
// ==========================================
service RAGService {
  // --- For Chat Flow ---
  // Retrieves relevant text chunks for a query
  rpc RetrieveContext (SearchRequest) returns (SearchResponse);

  // --- For Management Flow ---
  // Triggers the worker to process a file (Async)
  rpc TriggerSync (SyncRequest) returns (SyncResponse);
  
  // Deletes vectors for a specific document
  rpc DeleteVectors (DeleteVectorRequest) returns (DeleteVectorResponse);
  
  // Debug/Management: Check vectors for a doc
  rpc GetVectors (GetVectorRequest) returns (GetVectorResponse);
}

message SearchRequest {
  string query_text = 1;
  int32 top_k = 2;
}

message SearchResponse {
  // We reuse the top-level ContextChunk message here
  repeated ContextChunk chunks = 1;
}

message SyncRequest {
  string file_path = 1; // Path on shared volume
  string doc_id = 2;
}

message SyncResponse {
  string status = 1; // "Queued", "Processing"
  string job_id = 2;
}

message DeleteVectorRequest {
  string doc_id = 1;
}

message DeleteVectorResponse {
  bool success = 1;
}

message GetVectorRequest {
  string doc_id = 1;
}

message GetVectorResponse {
  int32 vector_count = 1;
}

// ==========================================
// 3. CHAT SERVICE (The Orchestrator)
// Used by: API Gateway (for /chat and /ws/chat)
// ==========================================
service ChatService {
  // Standard Text Chat: Takes query, returns answer + citations
  rpc Interact (ChatRequest) returns (ChatResponse);

  // Audio Streaming: Takes audio stream, returns text updates (transcription/answer)
  rpc StreamAudioChat (stream AudioChunk) returns (stream ChatStreamResponse);
}

message ChatRequest {
  string user_query = 1;
  string session_id = 2;
}

message ChatResponse {
  string text = 1;
  // Returns the context used to generate the answer
  repeated ContextChunk context_chunks = 2; 
}

message AudioChunk {
  bytes content = 1;          // The raw audio bytes (WebM/WAV)
  string session_id = 2;      // To track conversation context
  string mime_type = 3;       // e.g. "audio/webm"
}

message ChatStreamResponse {
  string text_chunk = 1;      // Partial text (transcription or answer)
  string event_type = 2;      // "transcription", "thinking", "answer", "done"
}